src/main/scala/PSO/PSOSystemFactory.scala

```diff
@@ -21,22 +21,24 @@ package pso
 import akka.actor.{ActorSystem, Props}
 
 class PSOSystemFactory(config: PSOConfiguration) {
-  val problemType = config.problemType
-  val solutions = config.initialSolutions
-  val topology = config.topology
+  private val problemType = config.problemType
+  private val solutions = config.initialSolutions
+  private val topology = config.topology
   private def obj = config.objectiveFunction
 
-  def build(randomSeed: Option[Int] = None): PSOSystem = {
-    val system = ActorSystem("SwarmSystem")
-    val master = system.actorOf(Props(
-      classOf[SwarmMaster],
-      obj,
-      config.problemType
-    ))
-    val r = randomSeed match {
-      case None => new scala.util.Random(System.currentTimeMillis)
-      case Some(seed) => new scala.util.Random(seed)
+  def build(randomSeed: Option[Int] = None,
+    actorSystem: Option[ActorSystem] = None): PSOSystem = {
+    val system = actorSystem match {
+      case None => ActorSystem("PSOSwarmSystem")
+      case Some(s) => s
     }
+    val master = system.actorOf(Props(new SwarmMaster(
+      // SwarmMaster requires objective function taking a List[AnyVal] as
+      // it is general to all optimization algorithms
+      X => obj(X.asInstanceOf[List[Double]]),
```

Comment:
This is a scary repercussion of using AnyVal - if something goes wrong and can't be case as a List[Double], it will throw an exception (not caught either).